solarforecastarbiter/pvmodel.py

```diff
@@ -0,0 +1,281 @@
+        system, apparent_zenith, azimuth)
+
+    dni_extra = pvlib.irradiance.get_extra_radiation(apparent_zenith.index)
+
+    poa_sky_diffuse = pvlib.irradiance.haydavies(
+        surface_tilt, surface_azimuth,
+        dhi, dni, dni_extra,
+        solar_zenith=apparent_zenith,
+        solar_azimuth=azimuth,
+        dni_extra=dni_extra, model='haydavies')
+    # assumes albedo = 0.25
+    poa_ground_diffuse = pvlib.irradiance.get_ground_diffuse(surface_tilt, ghi)
+    aoi_modifier = pvlib.pvsystem.physicaliam(aoi)
+    beam_component = dni * pvlib.tools.cosd(aoi)
+    beam_component = beam_component.clip(lower=0) * aoi_modifier
+    poa_effective = beam_component + poa_sky_diffuse + poa_ground_diffuse
+    return poa_effective
+
+
+def calculate_power(system, poa_effective, temp_air=20, wind_speed=1):
+    """
+    Calcuate AC power from system metadata, plane of array irradiance,
+    and weather data using the PVWatts model.
+
+    Parameters
+    ----------
+    system : datamodel.SolarPowerPlant
+    poa_effective : pd.Series
+    temp_air : pd.Series
+    wind_speed : pd.Series
+
+    Returns
+    -------
+    ac_power : pd.Series
+    """
+    pvtemps = pvlib.pvsystem.sapm_celltemp(poa_effective,
+                                           wind_speed, temp_air)
+    dc = pvlib.pvsystem.pvwatts_dc(poa_effective, pvtemps['temp_cell'],
+                                   system.modeling_parameters.dc_capacity)
+    dc *= (1 - system.modeling_parameters.dc_loss_factor / 100)
+    ac = pvlib.pvsystem.pvwatts_ac(dc, system.modeling_parameters.dc_capacity)
+    ac = ac.clip(upper=system.modeling_parameters.ac_capacity)
+    ac *= (1 - system.modeling_parameters.ac_loss_factor / 100)
+    return ac
+
+
+def ghi_to_power(system, ghi, temp_air=20, wind_speed=1):
+    """
+    Calcuate AC power from system metadata and ghi.
+
+    Parameters
+    ----------
+    system : datamodel.SolarPowerPlant
+    ghi : pd.Series
+    temp_air : pd.Series
+    wind_speed : pd.Series
+
+    Returns
+    -------
+    ac_power : pd.Series
+    """
+    solar_position = calculate_solar_position(system, ghi.index)
+    dni, dhi = complete_irradiance_components(ghi, solar_position['zenith'])
+    poa_effective = calculate_poa_effective(
+        system, solar_position['apparent_zenith'], solar_position['azimuth'],
+        ghi, dni, dhi)
+    ac = calculate_power(system, poa_effective, temp_air=temp_air,
+                         wind_speed=wind_speed)
+    return ac
+
+
+def ghi_dni_dhi_to_power(system, ghi, dni, dhi, temp_air=20, wind_speed=1):
+    """
+    Calcuate AC power from system metadata and ghi, dni, dhi.
+
+    Parameters
+    ----------
+    system : datamodel.SolarPowerPlant
+    ghi : pd.Series
+    dni : pd.Series
+    dhi : pd.Series
+    temp_air : pd.Series
+    wind_speed : pd.Series
+
+    Returns
+    -------
+    ac_power : pd.Series
+    """
+    solar_position = calculate_solar_position(system, ghi.index)
```

Comment:
this goes against calculating solar_position once and using it repeatedly

Reply:
was meant to be an example, but whatever